reco_utils/recommender/sar/sar_singlenode.py

```diff
@@ -50,533 +27,329 @@ class SARSingleNode:
+
+        return sparse.coo_matrix(
+            (df[self.col_rating], (df[self.col_user_id], df[self.col_item_id])),
+            shape=(n_users, n_items),
+        ).tocsr()
 
     def compute_coocurrence_matrix(self, df, n_users, n_items):
-        """ Coocurrence matrix
+        """ Co-occurrence matrix
         C = U'.transpose() * U'
         where U' is the user_affinity matrix with 1's as values (instead of ratings).
+
         Args:
-            df (pd.DataFrame): Hashed df of users and items.
+            df (pd.DataFrame): Indexed df of users and items.
             n_users (int): Number of users.
             n_items (int): Number of items.
         Returns:
-            np.array: Coocurrence matrix
+            np.array: Co-occurrence matrix
         """
-        self.time()
-        float_type = df[self.col_rating].dtype
+
         user_item_hits = (
             sparse.coo_matrix(
                 (
-                    np.array([1.0] * len(df[self._col_hashed_users])).astype(float_type),
-                    (df[self._col_hashed_users], df[self._col_hashed_items]),
+                    np.repeat(1, df.shape[0]),
+                    (df[self.col_user_id], df[self.col_item_id]),
                 ),
-                shape=(n_users, n_items)
+                shape=(n_users, n_items),
             )
-                .todok()
-                .tocsr()
+            .tocsr()
+            .astype(df[self.col_rating].dtype)
         )
 
         item_cooccurrence = user_item_hits.transpose().dot(user_item_hits)
-
-        if self.debug:
-            cnt = df.shape[0]
-            elapsed_time = self.time()
-            self.timer_log += [
-                "Item cooccurrence calculation:\t%d\trows in\t%s\tseconds -\t%f\trows per second."
-                % (cnt, elapsed_time, float(cnt) / elapsed_time)
-            ]
-
-        self.time()
         item_cooccurrence = item_cooccurrence.multiply(
             item_cooccurrence >= self.threshold
         )
-        if self.debug:
-            elapsed_time = self.time()
-            self.timer_log += [
-                "Applying threshold:\t%d\trows in\t%s\tseconds -\t%f\trows per second."
-                % (cnt, elapsed_time, float(cnt) / elapsed_time)
-            ]
+
         return item_cooccurrence
 
     def fit(self, df):
-        """Main fit method for SAR"""
+        """Main fit method for SAR
 
-        log.info("Collecting user affinity matrix...")
-        self.time()
-        # use the same floating type for the computations as input
-        float_type = df[self.col_rating].dtype
-        if not np.issubdtype(float_type, np.floating):
-            raise ValueError(
-                "Only floating point data types are accepted for the rating column. Data type was {} "
-                "instead.".format(float_type)
-            )
+        Args:
+            df (pd.DataFrame): User item rating dataframe
+        """
+
+        # Map a continuous index to user / item ids
```

Comment:
hey @gramhagen can we refactor the data indexing in a different method from fit? I need this for having consistency with the other algos when doing the benchmark https://github.com/Microsoft/Recommenders/issues/464. Ideally, I would need something like this:
```
index_df = index_data(df)
model = SAR()
model.fit(index_df)
```
or  in case the index needs to maintain state from SAR
```
model = SAR()
index_df = model.index_data(df)
model.fit(index_df)
```

if you think this will take a long time or if you don't have time for doing this, we can add an issue and any of us can make this work on a different PR.

Reply:
what are you using index_df for outside the model?

Reply:
I'm timing all the training method and for the rest of the algos, I'm not taking into account the data transformations they make before training. Here you can see a partial work https://github.com/Microsoft/Recommenders/blob/benchmark_feb/notebooks/03_evaluate/comparison_movielens.ipynb

Reply:
Hmm, I think it's fair to consider this as part of the algorithm since it is required for training / prediction. Separating it into another method just adds an unnecessary step for the user to remember.

Reply:
k, moved this into a separate method (set_index) which is called by fit if it hasn't been called externally, so if you want to time it without including the indexing (which is a pretty small amount of time relatively) you can. And if it's not called fit will take care of it. Everybody wins! =)

Reply:
that idea is really great.

Reply:
I don't think we use indexing in an algorithm other than SAR. But we do require data in different formats e.g. DeepRec accepts LIBFFM format. I agree with @gramhagen that this indexing is internal and can be hidden from the user. Otherwise, if we think it should be part of the data format and other algos will use it, we should make it available outside the scope of SAR, as a dataset util.